File: app/src/main/java/lsafer/edgeseek/legacy/SingleToast.java
/*
 *	Copyright 2020 LSafer
 *
 *	Licensed under the Apache License, Version 2.0 (the "License");
 *	you may not use this file except in compliance with the License.
 *	You may obtain a copy of the License at
 *
 *	    http://www.apache.org/licenses/LICENSE-2.0
 *
 *	Unless required by applicable law or agreed to in writing, software
 *	distributed under the License is distributed on an "AS IS" BASIS,
 *	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *	See the License for the specific language governing permissions and
 *	limitations under the License.
 */
package lsafer.edgeseek.legacy;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import java.util.Objects;

import lsafer.edgeseek.App;
import lsafer.edgeseek.R;

/**
 * Two toasts of this class can't be displayed at the same time.
 *
 * @author lsafer
 * @version 0.1.5
 * @since 08-Oct-2019
 */
final public class SingleToast extends Toast {
	/**
	 * The current displaying toast.
	 */
	private static SingleToast current;

	/**
	 * Constructs a new toast.
	 *
	 * @param context to be used
	 * @throws NullPointerException if the given context is null
	 */
	public SingleToast(Context context) {
		super(context);
	}

	/**
	 * relevant to the method {@link Toast#makeText(Context, CharSequence, int)}.
	 *
	 * @param context  to be used
	 * @param string   to be displayed
	 * @param duration to show the toast within
	 * @return the toast constructed
	 * @throws NullPointerException if the given 'context' or 'string' is null
	 */
	public static SingleToast makeText(Context context, String string, int duration) {
		Objects.requireNonNull(context, "context");
		Objects.requireNonNull(string, "string");

		context.setTheme(App.data.getTheme());
		SingleToast toast = new SingleToast(context);

		toast.setDuration(duration);
		View view = LayoutInflater.from(context).inflate(R.layout.legacy_toast, null);
		view.<TextView>findViewById(R.id.txt).setText(string);

		toast.setView(view);
		return toast;
	}

	@Override
	public void show() {
		synchronized (SingleToast.class) {
			if (SingleToast.current != null)
				SingleToast.current.cancel();
			SingleToast.current = this;
			super.show();
		}
	}
}

File: app/src/main/kotlin/net/lsafer/edgeseek/component/SingleToast.kt
/*
 *	Copyright 2020 LSafer
 *
 *	Licensed under the Apache License, Version 2.0 (the "License");
 *	you may not use this file except in compliance with the License.
 *	You may obtain a copy of the License at
 *
 *	    http://www.apache.org/licenses/LICENSE-2.0
 *
 *	Unless required by applicable law or agreed to in writing, software
 *	distributed under the License is distributed on an "AS IS" BASIS,
 *	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *	See the License for the specific language governing permissions and
 *	limitations under the License.
 */

@file:SuppressLint("StaticFieldLeak")

package net.lsafer.edgeseek.component

import android.annotation.SuppressLint
import android.content.Context
import android.os.Build
import android.util.Log
import android.view.Gravity
import android.view.WindowManager
import android.view.WindowManager.LayoutParams
import android.widget.TextView
import androidx.cardview.widget.CardView
import kotlinx.coroutines.*

object SingleToast {
    lateinit var text: String
    lateinit var cardView: CardView
    lateinit var textView: TextView
    var showId = 0

    var params = LayoutParams().also {
        @Suppress("DEPRECATION")
        it.type = when {
            Build.VERSION.SDK_INT >= 26 ->
                LayoutParams.TYPE_APPLICATION_OVERLAY
            else ->
                LayoutParams.TYPE_PHONE
        }
        @Suppress("DEPRECATION")
        it.flags = LayoutParams.FLAG_NOT_FOCUSABLE or
                LayoutParams.FLAG_LAYOUT_IN_SCREEN or
                LayoutParams.FLAG_LAYOUT_NO_LIMITS or
                LayoutParams.FLAG_NOT_TOUCH_MODAL or
                LayoutParams.FLAG_SHOW_WHEN_LOCKED
        it.gravity = Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL
        it.height = 100
        it.width = 200
        it.y = 50
    }
    var attached: Boolean = false

    fun display(context: Context, text: String) {
        buildView(context)
        updateText(text)
        show(context)
    }

    fun show(context: Context) {
        attachToWindow(context)

        val expectedId = ++showId

        CoroutineScope(Dispatchers.Unconfined).launch {
            delay(2_000)
            if (showId == expectedId) {
                withContext(Dispatchers.Main) {
                    detachFromWindow(context)
                }
            }
        }
    }

    fun buildView(context: Context) {
        if (!SingleToast::textView.isInitialized || !SingleToast::cardView.isInitialized) {
            cardView = CardView(context)
            textView = TextView(context)
            textView.gravity = Gravity.CENTER
            cardView.radius = 25f
            cardView.addView(textView)
        }
    }

    fun updateText(newText: String) {
        textView.text = newText
    }

    fun attachToWindow(context: Context) {
        val manager = context.getSystemService(WindowManager::class.java)

        try {
            if (!attached) {
                manager.addView(cardView, params)
                attached = true
            }
        } catch (e: Exception) {
            Log.e("LALA_LAND", "attachToWindow: ", e)
        }
    }

    fun detachFromWindow(context: Context) {
        val manager = context.getSystemService(WindowManager::class.java)

        try {
            if (attached) {
                manager.removeView(cardView)
                attached = false
            }
        } catch (e: Exception) {
            Log.e("LALA_LAND", "detachFromWindow: ", e)
        }
    }
}
